R/generalized_normal.R

```diff
@@ -0,0 +1,188 @@
+#' @param beta shape, beta > 0
+#'
+#' @return Random numbers from the generalized_normal distribution.
+#' @export
+#'
+#' @examples hist(rgeneralized_normal(100, mu = 2, sigma = 2, beta = 2))
+rgeneralized_normal <- function(n, mu = 0, sigma = 1, beta = 1) {
+  # check the arguments
+  if (!isNat_len(n)) {
+    stop("The number RNG-samples has to be a scalar natural")
+  }
+  return(qgeneralized_normal(p = runif(n, min = 0, max = 1), mu = mu, sigma = sigma, beta = beta))
+}
+
+#' Log-Likelihood of the generalized_normal distribution
+#'
+#' @param i BRMS indices
+#' @param prep BRMS data
+#'
+#' @return Log-Likelihood of generalized_normal given data in prep
+log_lik_generalized_normal <- function(i, prep) {
+  mu <- brms::get_dpar(prep, "mu", i = i)
+  sigma <- brms::get_dpar(prep, "sigma", i = i)
+  beta <- brms::get_dpar(prep, "beta", i = i)
+  y <- prep$data$Y[i]
+  return(dgeneralized_normal(x = y, mu = mu, sigma = sigma, beta = beta, log = TRUE))
+}
+
+
+#' posterior_predict for the generalized_normal distribution
+#'
+#' @param i BRMS indices
+#' @param prep BRMS data
+#' @param ... catchall argument
+#'
+#' @return Draws from the Posterior Predictive Distribution
+posterior_predict_generalized_normal <- function(i, prep, ...) {
+  mu <- brms::get_dpar(prep, "mu", i = i)
+  sigma <- brms::get_dpar(prep, "sigma", i = i)
+  beta <- brms::get_dpar(prep, "beta", i = i)
+  return(rgeneralized_normal(n = prep$ndraws, mu = mu, sigma = sigma, beta = beta))
+}
+
+#' posterior_epred for the generalized_normal distribution
+#'
+#' @param prep BRMS data
+#'
+#' @return Expected Values of the Posterior Predictive Distribution
+posterior_epred_generalized_normal <- function(prep) {
+  return(brms::get_dpar(prep, "mu"))
+}
+
+#' Generalized Normal BRMS family
+#'
+#' @param link Link function for function
+#' @param link_sigma Link function for sigma argument
+#' @param link_b Link function for beta argument
+#'
+#' @return BRMS generalized_normal distribution family
+#' @export
+#'
+#' @examples data <- list(y = rgeneralized_normal(n = 1000, mu = 2, sigma = 2, beta = 4))
+#' fit <- brms::brm(formula = y ~ 1, data = data,
+#'   family = generalized_normal(), stanvars = generalized_normal()$stanvars,
+#'   init = 0.1)
+#' plot(fit)
+generalized_normal <- function(link = "identity", link_sigma = "log", link_b = "log") {
+  family <- brms::custom_family(
+    "generalized_normal",
+    dpars = c("mu", "sigma", "beta"),
+    links = c(link, link_sigma, link_b),
+    lb = c(-NA, 0, 0),
+    ub = c(NA, NA, NA),
+    type = "real",
+    log_lik = log_lik_generalized_normal,
+    posterior_predict = posterior_predict_generalized_normal,
+    posterior_epred = posterior_epred_generalized_normal
+  )
+  family$stanvars <- brms::stanvar(
+    scode = "
+      real generalized_normal_lpdf(real y, real mu, real sigma, real beta) {
+        return log(beta) - (log(2) + log(sigma) + lgamma(1/beta))
+          - (abs(y - mu) / sigma)^beta;
```

Comment:
please doublecheck that the R and stan code are the same here. While you don't need the second abs above, it is missing here.

Reply:
Removed redundant abs in R, so R and Stan are identical.